src/OptimizedJsonRepository.php

```diff
@@ -0,0 +1,222 @@
+use Webmozart\Glob\Glob;
+use Webmozart\PathUtil\Path;
+
+/**
+ * A repository backed by a JSON file optimized for reading.
+ *
+ * The generated JSON file is described by res/schema/repository-schema-1.0.json.
+ *
+ * Resources can be added with the method {@link add()}:
+ *
+ * ```php
+ * use Puli\Repository\JsonRepository;
+ *
+ * $repo = new JsonRepository('/path/to/repository.json', '/path/to/project');
+ * $repo->add('/css', new DirectoryResource('/path/to/project/res/css'));
+ * ```
+ *
+ * When adding a resource, the added filesystem path is stored in the JSON file
+ * under the key of the Puli path. The path is stored relatively to the base
+ * directory passed to the constructor. Directories will be expanded and all
+ * nested files will be added to the mapping file as well:
+ *
+ * ```json
+ * {
+ *     "/css": "res/css",
+ *     "/css/style.css": "res/css/style.css"
+ * }
+ * ```
+ *
+ * Mapped resources can be read with the method {@link get()}:
+ *
+ * ```php
+ * $cssPath = $repo->get('/css')->getFilesystemPath();
+ * ```
+ *
+ * You can also access nested files:
+ *
+ * ```php
+ * echo $repo->get('/css/style.css')->getBody();
+ * ```
+ *
+ * Since nested files are searched during {@link add()} and added to the JSON
+ * file, this repository does not detect any files that you add to a directory
+ * after adding that directory to the repository. This means that accessing
+ * files is very fast, but also that the usage of this repository implementation
+ * can be cumbersome in development environments. There you are recommended to
+ * use {@link JsonRepository} instead.
+ *
+ * @since  1.0
+ *
+ * @author Bernhard Schussek <bschussek@gmail.com>
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class OptimizedJsonRepository extends AbstractJsonRepository implements EditableRepository
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function insertReference($path, $reference)
+    {
+        $this->json[$path] = $reference;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function removeReferences($glob)
+    {
+        $removed = 0;
+
+        foreach ($this->getReferencesForGlob($glob.'{,/**/*}') as $path => $reference) {
+            ++$removed;
+
+            unset($this->json[$path]);
+        }
+
+        return $removed;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getReferencesForPath($path)
+    {
+        if (!array_key_exists($path, $this->json)) {
+            return array();
+        }
+
+        $reference = $this->json[$path];
+
+        // We're only interested in the first entry of eventual arrays
+        if (is_array($reference)) {
+            $reference = reset($reference);
+        }
+
+        if ($this->isFilesystemReference($reference)) {
+            $reference = Path::makeAbsolute($reference, $this->baseDirectory);
+
+            // Ignore non-existing files. Not sure this is the right
+            // thing to do.
```

Comment:
I logged a warning now in this case. I don't want to fail completely because then find() will fail even if just one reference is configured with a wrong path. Since references could be configured in other packages that the user has no control over, failing completely is not very user friendly.